docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,608 @@
+  when a `child-3` is added, it will be covered with the existing subscription itself.
+  But, if these child devices have nested child devices, the parent will have to subscribe to,
+  `tedge/child-1/+/#` and `tedge/child-2/+/#` separately,
+  and when `child-3` is added, dynamically subscribe to `tedge/child-3/+/#` as well.
+
+## Proposal 2: Unified topics for every "thing"
+
+This proposal is built on top of the following assumptions:
+1. Every entity, a child device, descendent child devices or services, must register with thin-edge
+   to get their unique IDs generated, unless these entities can ensure uniqueness across the entire fleet.
+1. No dynamic registration for anything.
+1. Avoiding the entity type(device or service, child or descendent etc) from the topics reduces the payload size as well,
+   as the type of an entity needs to be declared only once and need not be repeated over and over for every message.
+
+The proposal is as follows:
+
+A topic scheme that doesn't differentiate between parent child devices or services and
+does not declare the parent-child relation in the topics either, as follows:
+
+```
+tedge/<entity-id>/...
+```
+
+...where `id` could be the `id` of the main device, child device or a service on any of them.
+The alias `main` can still be used for the main tedge device.
+
+Examples:
+* Main device measurements: `tedge/<tedge-device-id>/telemetry/measurements`
+* Main device service measurements: `tedge/<service-id>/telemetry/measurements`
+* Child device measurements: `tedge/<child-device-id>/telemetry/measurements`
+* Child device service measurements: `tedge/<child-service-id>/telemetry/measurements`
+
+```admonish note
+`tedge/main` could be just used as an alias for `tedge/<tedge-device-id>` for simplicity.
+```
+
+**Pros**
+
+1. Due to the lack of distinction in the topics between parent devices and child devices,
+   it is easier to write code for "the device" irrespective of whether it is deployed as a parent or child.
+1. Minimal payload size as we're only including the `id`s everywhere without repeating any context info.
+
+**Cons**
+
+1. Not easy to differentiate the context(from parent or child) easily from the message.
+1. The `id`s must be unique between all devices and services in a deployment.
+1. Not easy to do subscriptions like: "measurements only from child devices or only from services, excluding parent",
+   without keeping track of all the `id`s of services registered with that child device.
+   Easy wild card subscriptions are not possible at all.
+1. No automatic registration as bootstrapping is mandatory for everything.
+
+## Enhancements
+
+* Introduce a `tedge/self` topic that can be used by developers to write context-agnostic tedge components,
+  without worrying about device IDs and their context (whether deployed on main or child device).
+  The `tedge/self` prefix must be mapped to `tedge/main` on the main device and `tedge/<child-id>` on the child devices.
```

Comment:
You mean using a rewriting rule applied to the topic name before the message is published?  Or do you imagine independent mosquitto instances connected by local bridges?

Reply:
I was imagining a chain of devices, each having its own broker, bridged with its immediate parent. Something like:

```mermaid
graph LR
    tedge-broker --> child-broker
    child-broker --> grand-child-broker
```

The `tedge connect` command on the `tedge` device will create the mapping rule from `tedge/self` to `tedge/main` on the tedge-broker. Similarly, the "connection mechanism" on `child` will create the mapping from `tedge/self` to `tedge/child` and so on.

The local processes running on the device, generating data on behalf of that device, can continue publishing to `tedge/self` topics, which will be mapped to their respective IDs, by the broker bridge config, before they are sent to their parent's broker.

Reply:
No longer relevant. Proposal 3 was chosen which supports registration message which indicate its parent. The id of the device can also be decoupled from the topic structure.